src/optimization/bounds.hpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Sleipnir contributors
+
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <span>
+#include <tuple>
+#include <utility>
+
+#include <Eigen/Core>
+#include <Eigen/SparseCore>
+
+#include "sleipnir/autodiff/expression_type.hpp"
+#include "sleipnir/autodiff/variable.hpp"
+#include "sleipnir/util/assert.hpp"
+#include "sleipnir/util/small_vector.hpp"
+
+// See docs/algorithms.md#Works_cited for citation definitions
+
+namespace slp {
+
+/**
+ * A "bound constraint" is any linear constraint in one scalar variable.
+ * Computes which constraints, if any, are bound constraints, whether or not
+ * they're feasible (given previously encountered bounds), and the tightest
+ * bounds on each decision variable.
+ *
+ * @param decisionVariables Decision variables corresponding to each column of
+ *   A_i.
+ * @param inequalityConstraints Variables representing the left-hand side of
+ *   cᵢ(decisionVariables) ≥ 0.
+ * @param A_i The Jacobian of inequalityConstraints wrt decisionVariables,
+ *   evaluated anywhere, in *row-major* storage; in practice, since we typically
+ *   store Jacobians column-major, the user of this function must perform a
+ *   transpose.
+ */
+inline std::tuple<Eigen::ArrayX<bool>, small_vector<std::pair<double, double>>,
+                  small_vector<std::pair<Eigen::Index, Eigen::Index>>>
+get_bounds(const std::span<Variable> decision_variables,
```

Comment:
If you meant a range of immutable elements, you want `std::span<const Variable>` instead.

Reply:
I did not; I just wanted to make it impossible to re-assign the span itself (although perhaps this is needlessly paranoid).

Reply:
A `const span` argument is unusual in the same way a `const int` argument is. Since it's taken by value, the function body modifying it doesn't affect the original view.